builtin/providers/google/resource_compute_instance.go

```diff
@@ -791,24 +851,48 @@ func resourceComputeInstanceRead(d *schema.ResourceData, meta interface{}) error
 		d.Set("tags_fingerprint", instance.Tags.Fingerprint)
 	}
 
-	disks := make([]map[string]interface{}, 0, 1)
+	disksCount := d.Get("disk.#").(int)
+	attachedDisksCount := d.Get("attached_disk.#").(int)
+	disks := make([]map[string]interface{}, 0, disksCount)
+	attached_disks := make([]map[string]interface{}, 0, attachedDisksCount)
+
+	if expectedDisks := disksCount + attachedDisksCount; len(instance.Disks) != expectedDisks {
+		return fmt.Errorf("Expected %d disks, API returned %d", expectedDisks, len(instance.Disks))
+	}
+
+	// This loop takes advanatage of the fact that disks are returned in the
+	// order they're created, and we create disks that are part of "disk" before
+	// ones that are part of "attached_disk"
 	for i, disk := range instance.Disks {
-		di := map[string]interface{}{
-			"disk":                    d.Get(fmt.Sprintf("disk.%d.disk", i)),
-			"image":                   d.Get(fmt.Sprintf("disk.%d.image", i)),
-			"type":                    d.Get(fmt.Sprintf("disk.%d.type", i)),
-			"scratch":                 d.Get(fmt.Sprintf("disk.%d.scratch", i)),
-			"auto_delete":             d.Get(fmt.Sprintf("disk.%d.auto_delete", i)),
-			"size":                    d.Get(fmt.Sprintf("disk.%d.size", i)),
-			"device_name":             d.Get(fmt.Sprintf("disk.%d.device_name", i)),
-			"disk_encryption_key_raw": d.Get(fmt.Sprintf("disk.%d.disk_encryption_key_raw", i)),
-		}
-		if disk.DiskEncryptionKey != nil && disk.DiskEncryptionKey.Sha256 != "" {
-			di["disk_encryption_key_sha256"] = disk.DiskEncryptionKey.Sha256
+		if i < disksCount {
```

Comment:
Done.

(technically we're still relying on the fact they're returned in the same order by mixing disk.[property] and d.Get([property]) calls, but your way does certainly look less hacky)